fbinternal/teravm/fabfile.py

```diff
@@ -0,0 +1,654 @@
+    the remote host. If empty file name is passed, password-based ssh will
+    work instead. This can be used if the script is run manually.
+
+    test_file: a test file to use instead of the s6a and gxgy defaults.
+    """
+    if isinstance(test_files, str):
+        test_files = [test_files]
+    test_output = []
+
+    for test_file in test_files:
+        fastprint("Run test for file %s\n" % (test_file))
+        _setup_env("ng40", VM_IP_MAP[setup]["ng40"], key_filename)
+        with cd("/home/ng40/magma/automation"):
+            with hide("warnings", "running", "stdout"), settings(warn_only=True):
+                output = run("ng40test %s" % test_file)
+                test_output.append(output)
+        fastprint("Done with file %s\n" % (test_file))
+
+    verdicts = _parse_stats(test_output)
+    return verdicts
+
+
+def _set_magma_apt_repo():
+    err = None
+    with settings(abort_exception=FabricException):
+        try:
+            # add repo to source file (same as add-apt-repo
+            repo_apt_string = "deb {} {}".format(AGW_APT_SOURCE, AGW_APT_BRANCH)
+            ignore_comments = "/^[[:space:]]*#/!"
+            sudo("touch {}".format(AGW_ATP_FILE))
+            # Replace non commented lines with the wrong repo, or add it if missing
+            sudo(
+                "grep -q '{source}' {sFile} && "
+                "sed -i '{ign_com}s,.*{source}.*,{repo},g' {sFile} || "
+                "echo '{repo}' >> {sFile} ".format(
+                    ign_com=ignore_comments,
+                    source=AGW_APT_SOURCE,
+                    repo=repo_apt_string,
+                    sFile=AGW_ATP_FILE,
+                )
+            )
+        except Exception:
+            err = "Error changing ATP repo\n"
+            fastprint(err)
+    return err
+
+
+def _parse_stats(teravm_raw_result):
+    """
+    Gets stats from teraVM result output string
+
+    teravm_test_result: output comming from the teravm stdout
+    """
+    verdicts = collections.defaultdict(list)
+
+    pattern = r"Verdict\((?P<test_case>\w+)\) = VERDICT_(?P<verdict>\w+)"
+    for fileResults in teravm_raw_result:
+        for line in fileResults.splitlines():
+            match = re.match(pattern, line)
+            if match:
+                verdict = match.groupdict()["verdict"]
+                verdicts[verdict].append(line)
+    return verdicts
+
+
+def _post_to_slack(setup, verdicts, hash=None):
+    """
+    Post test results to slack. If hash is given, also post on which build
+    hash the result is based on.
+
+    message: proper formated string with the message to be posted
+    hash: The hash of the build the system was upgraded to right before test.
+    """
+    if not verdicts:
+        fastprint("Verdicts are empty. Something went wrong running" "teravm tests.")
+        return
+    data = _compose_slack_message(setup, verdicts, hash)
+    _send_message_to_slack(data)
+
+
+def _send_message_to_slack(data):
```

Comment:
Hmm I would remove that code and only focus on upgrade and triggering features 
what do you think ?